draft-ietf-quic-transport.md

```diff
@@ -3381,51 +3382,39 @@ Header Form:
 
 : The most significant bit (0x80) of byte 0 is set to 0 for the short header.
 
-Key Phase Bit:
-
-: The second bit (0x40) of byte 0 indicates the key phase, which allows a
-  recipient of a packet to identify the packet protection keys that are used to
-  protect the packet.  See {{QUIC-TLS}} for details.
-
-\[\[Editor's Note: this section should be removed and the bit definitions
-changed before this draft goes to the IESG.]]
+Fixed Bit:
 
-Third Bit:
+: The next bit (0x40) of byte 0 is set to 1.  Packets containing a zero value
+  for this bit are not valid packets in this version and MUST be discarded.
 
-: The third bit (0x20) of byte 0 is set to 1.
+Spin Bit (S):
 
-\[\[Editor's Note: this section should be removed and the bit definitions
-changed before this draft goes to the IESG.]]
+: The sixth bit (0x20) of byte 0 is the Latency Spin Bit, set as described in
+  {{!SPIN=I-D.ietf-quic-spin-exp}}.
 
-Fourth Bit:
+Reserved Bits (R):
 
-: The fourth bit (0x10) of byte 0 is set to 1.
+: The next two bits (those with a mask of 0x18) of byte 0 are reserved.  These
+  bits are protected using header protection (see Section 5.4 of
+  {{QUIC-TLS}}).  The value included prior to protection MUST be set to 0.  An
```

Comment:
What important purpose is served by requiring the value of these bits to be 0?  This precludes the use of these bits for experimental purposes, like VEC and LOSS detection.

Before the use of these bits is standardized, the risk of middleboxes ossifying occasional experimental use of the bits seems less significant than the loss of opportunity to experiment.

Reply:
They're set to 0 prior to protection, which means a middlebox can't use them anyway unless you share the header protection key with it.

Reply:
The header protection is an XOR with a mask.  Setting these bits to a desired experimental value _after_ the header protection is applied (after XOR) works ok, if the other endpoint did not require the bits to be 0 after the header protection is removed.

Hence the question -- why a hard requirement to have these bits 0 after removing header protection, if the endpoint does not use these bits for anything?

Reply:
Because otherwise they become scratch space where a middlebox could place a persistent marking on the packets.  If peers want to discard integrity checks on some portion of the data they send/receive, they can certainly do so -- but not unilaterally.  By default, modification of the packet makes the packet invalid.

Reply:
Surely AEAD verification would prevent scratch spaces regardless of whether the value is checked or not?

But I can see that setting the value to 0 simplifies future extensions or experiments.

Reply:
Right. An AEAD verification would prevent modification by middleboxes -- a desirable property.  Insistence on the value being 0 prevents experimental use for the purposes that do not require two cooperating endpoints (like Loss bit).

Reply:
I see that as a feature, rather than a bug. We’re making it harder for one endpoint to expose potentially privacy sensitive information to third parties, if its peer isn’t consenting. Of course, we can never fully prevent that, but at least we can make it harder and increase the overhead for that.

Reply:
If you want to negotiate a new version, you can assign these bits however you choose in that version.  Or, in this version, you could negotiate an extension that sets these in different ways (though probably only in short headers in practice).  That extension could also decide to remove any encryption.  So yes, you could use the spare bits for a VEC, if both peers agreed that was a good idea.

Until that time, this is protecting those bits the best way we know how.

Reply:
@marten-seemann, as you said, it is impossible to prevent one party from disclosing anything about the connection to any third party. No shenanigans can make a dent in the amount or quality of info disclosed by a determined party. However, this setup can make it harder to troubleshoot network operations (if that involves unilaterally setting these bits for experimentation).

@martinthomson, yes, and the original comment was looking for a clarification of the threat that we are "protecting those bits the best way we know how".  Is it protecting the bits from an endpoint bent on disclosing some private information to a third party?  Or is it protecting from an endpoint accidentally exposing some information while trying to run an experiment (i.e. since one cannot be sure that everyone will run safe experiments, no one is allowed to run experiments unilaterally)? Or is it protecting from a middlebox ossifying some experimental usage of these bits? Or is it protecting from something else?

Reply:
@kazuho, the suggestion is not bits-on-the-wire are 0 (which is an ossification concern).

The suggestion is that bits can be anything and they are AEAD protected but not masked -- just like the latency spin bit.  They could be set randomly or not randomly -- up to the endpoint.  An endpoint can already disclose any and all info to anyone without peer's consent, and these bits are by far not the most convenient mechanism for such disclosures.

Do I understand it correctly that the worry is about an _accidental_ and not a _deliberate_ disclosure?  (I do not think a worry about ossification of some experimental use it warranted, do you?)

Reply:
@igorlord 
> The suggestion is that bits can be anything and they are AEAD protected but not masked -- just like the latency spin bit.

Thank you for the clarification. I understand that.

> They could be set randomly or not randomly -- up to the endpoint.

That means that most endpoints will be required to set them randomly, at least initially. The issue here is that you need to randomize the bits separately, rather than just relying on header protection to randomize the bits. It is an complication.

> An endpoint can already disclose any and all info to anyone without peer's consent, and these bits are by far not the most convenient mechanism for such disclosures.

If that is the case, why do we need to care about the bits? If the endpoints have an easier way to expose the information, can't they just be randomized using header protection?

Reply:
@kazuho 
> That means that endpoints (that do not have a way to "use" the bits) will be required to set them randomly. The issue here is that you need to randomize the bits separately, rather than just relying on header protection to randomize the bits. It is an unnecessary complexity.

A PRNG is already needed for the spin bit's anonimity set, and for its fixed per-conneciton value when not spinning. A tiny increment, then, compared to the weight of version deployment for anybody willing to experiment with these "reserved" (should be called "frozen") bits.

> > An endpoint can already disclose any and all info to anyone without peer's consent, and these bits are by far not the most convenient mechanism for such disclosures.
> 
> If that is the case, why do we need to care about the bits? If the endpoints have an easier way to expose the information, can't they just be randomized using header protection?

What Igor means, I think, is that the protection against unilateral disclosure of *anything* is negligible (for one, the endpoint could send unencrypted copies of the entire flow somewhere else). While using the reserved bits (which were designed for exactly this) is clearly the most efficient way for on-path signalling (no overhead ; same-tuple ; sync with flow).

Reply:
@kazuho
> > An endpoint can already disclose any and all info to anyone without peer's consent, and these bits are by far not the most convenient mechanism for such disclosures.
> 
> If that is the case, why do we need to care about the bits? If the endpoints have an easier way to expose the information, can't they just be randomized using header protection?

An endpoint _deliberately_ trying to compromise the connection's privacy can do it easily (and will likely use a more structured info channel for that purpose). However, an endpoint that is _not_ trying to leak information and is just tagging packets for VEC or Loss detection or other operational purpose would find these bits a lot more convenient.  That said, I understand the concern about some inadvertent disclosure by a poorly designed experiment (if this is the concern).

Reply:
@ferrieux
> A PRNG is already needed for the spin bit's anonimity set, and for its fixed per-conneciton value when not spinning. A tiny increment, then, compared to the weight of version deployment for anybody willing to experiment with these "reserved" (should be called "frozen") bits.

Runtime overhead is the issue here; I do not think requiring endpoints to call CSPRNG for every packet they send is a good idea (note: use of non-cryptographically secure RNG is a way to leak PN). Re providing the freedom to "experiment", please see my comment below.

> While using the reserved bits (which were designed for exactly this) is clearly the most efficient way for on-path signalling (no overhead ; same-tuple ; sync with flow).

While I do not think reserved bits are "designed" for on-path signaling, I agree that it is a convenient way for exposing signals.

The trade-off here is if we want to pay the extra cost of calling CSPRNG for every packet we send, to provide the endpoints to use the bits exposing information to on-path devices without peers' consent.

@igorlord
> However, an endpoint that is not trying to leak information and is just tagging packets for VEC or Loss detection or other operational purpose would find these bits a lot more convenient. That said, I understand the concern about some inadvertent disclosure by a poorly designed experiment (if this is the concern).

Yes. That is exactly the concern. Remember that we were uncertain about the privacy concerns of the spin bit for very long time (IIRC the risk of exposing the latency of the hidden path behind a VPN was brought up pretty recently).

I think that we should allow endpoints to use the bits for exposing information only when  extensive privacy analysis has been done and the WG is convinced. Protecting the bits under header protection is the easiest way to guarantee that.

Reply:
I do not have experience operating a network with completely opaque traffic.  So it is very possible that experiments with these bits would not materialize even if the bits were available for experimentation.  On the other hand, it is possible that some experiments could prove valuable to operators and not be a privacy threat.

Reply:
To be clear, if client and server both agree to another use for these bits, they can be used for any purpose.  It's only unilateral experimentation that is prevented by the current arrangement, which - as other have said - is intentional.

Reply:
@martinthomson 
Sorry, the thread did get too unwieldy, indeed.  The specific change:

> The suggestion is that reserved bits can be anything and they are AEAD protected but not masked -- just like the latency spin bit.  If not in-use for a specific purpose by the sender, the bits should be randomized.

Reply:
Ack, thanks.

I'd suggest that you take this to the mailing list, and open an issue.  This is one of the worst possible venues for a discussion of this type.

Reply:
Ack. an issue for this makes sense,

Reply:
This conversation is getting carried away. @igorlord: If the bits can be anything, and if middleboxes start using those bits as scratch space, they're effectively ossified. Endpoints can't use them anymore, since middleboxes might stomp on them in the network. How do we counter that?

Reply:
>  How do we counter that?

@janaiyengar as I wrote in an earlier comment, the fields are protected as authenticated data in the AEAD tag, so middleboxes cannot change the bits without getting the packet rejected. At best they can read the bits iff they are outside header protection, but they are not.

Reply:
@mikkelfj : *facepalm* Right.

Reply:
Opened issue #2022.

Reply:
I think the spec is right. In the absence of standard, the use case for VEC or ERR in V1 cannot be "manage my network". De facto, it has to be "experiment with new algorithms that would let me manage my network better". And experimenters can in fact do that, if clients and servers cooperate.